src/dkg.rs

```diff
@@ -0,0 +1,413 @@
+            }
+            Self::MalformedParticipantProof(err) => {
+                write!(
+                    formatter,
+                    "provided proof of possession or public polynomial is malformed: {err}"
+                )
+            }
+            Self::InconsistentPublicShares(err) => {
+                write!(
+                    formatter,
+                    "public shares obtained from cumulated public polynomial are inconsistent: {err}"
+                )
+            }
+        }
+    }
+}
+
+/// Structure for collecting commitments of public share in committed Pedersen's
+/// distributed key generation.
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(bound = ""))]
+pub struct DkgCommitmentsCollector<G: Group> {
+    params: Params,
+    index: usize,
+    dealer: Dealer<G>,
+    commitments: Vec<Option<[u8; 32]>>,
+    decommitment: [u8; 32],
+}
+
+impl<G: Group> DkgCommitmentsCollector<G> {
+    /// Instantiates a distributed key generation participant.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index` is greater or equal to number of shares.
+    pub fn new<R: CryptoRng + RngCore>(params: Params, index: usize, rng: &mut R) -> Self {
+        assert!(index < params.shares);
+        let mut decommitment = [0u8; 32];
+        rng.fill_bytes(&mut decommitment);
+        Self {
+            params,
+            index,
+            dealer: Dealer::new(params, rng),
+            commitments: vec![None; params.shares],
+            decommitment,
+        }
+    }
+
+    /// Generates commitment of participant's share of a public key.
+    pub fn commitment(&mut self) -> [u8; 32] {
+        if let Some(c) = self.commitments[self.index] {
+            c
+        } else {
+            let (public_poly, _) = self.dealer.public_info();
+            let mut hasher = Sha256::new();
+            let mut bytes = vec![0_u8; G::ELEMENT_SIZE];
+            G::serialize_element(&public_poly[0], &mut bytes);
+            hasher.update(&bytes);
+            hasher.update(self.decommitment);
+            let c = hasher.finalize().into();
+            self.commitments[self.index] = Some(c);
+            c
+        }
+    }
+
+    /// Inserts commitment from other participant with index `participant_index`.
+    pub fn insert_commitment(&mut self, participant_index: usize, commitment: [u8; 32]) {
+        self.commitments[participant_index] = Some(commitment);
+    }
+
+    /// Returns indices of parties whose commitments were not provided.
+    pub fn missing_commitments(&self) -> Vec<usize> {
+        self.commitments
+            .iter()
+            .enumerate()
+            .filter_map(|(i, x)| if x.is_none() { Some(i) } else { None })
+            .collect()
+    }
+
+    /// If all commiements were received returns DkgShareCollector
+    /// otherwise returns None.
+    pub fn finish_commitment_phase(&self) -> Option<DkgSharesCollector<G>> {
```

Comment:
Since it's easy to check whether the operation will succeed beforehand using `Self::missing_commitments()`, I'd make this operation return a non-optional value and panic if any commitments are missing.